mbf_costmap_nav/src/mbf_costmap_nav/search_helper.cpp

```diff
@@ -0,0 +1,294 @@
+// mbf_costmap_nav
+#include "mbf_costmap_nav/search_helper.h"
+
+// std
+#include <queue>
+#include <unordered_set>
+
+namespace mbf_costmap_nav
+{
+
+SearchHelper::SearchHelper(const costmap_2d::Costmap2DROS* costmap, const SearchConfig& config,
+                           const std::optional<std::function<bool(const Cell, const Cell)>>& compare_strategy,
+                           const std::optional<SearchHelperViz>& viz)
+  : costmap_(costmap), config_(config), viz_(viz)
+{
+  Cell start;
+  costmap_->getCostmap()->worldToMap(config_.goal.x, config_.goal.y, start.x, start.y);
+  ROS_DEBUG_STREAM("Start cell: " << start.x << ", " << start.y);
+  EuclideanCompare euclidean_compare(start);
+  compare_strategy_ = std::move(compare_strategy.value_or(euclidean_compare));
+}
+
+std::vector<Cell> SearchHelper::getNeighbors(const costmap_2d::Costmap2D* costmap_2d, const Cell& cell)
+{
+  std::vector<Cell> neighbors;
+  neighbors.reserve(8);
+  for (int dx = -1; dx <= 1; ++dx)
+  {
+    for (int dy = -1; dy <= 1; ++dy)
+    {
+      if (dx == 0 && dy == 0)
+      {
+        continue;
+      }
+      unsigned int x = cell.x + dx;
+      unsigned int y = cell.y + dy;
+      if (x < costmap_2d->getSizeInCellsX() && y < costmap_2d->getSizeInCellsY() && x >= 0 && y >= 0)
+      {
+        neighbors.push_back(Cell{ x, y, costmap_2d->getCost(x, y) });
+      }
+    }
+  }
+  return neighbors;
+}
+
+std::vector<geometry_msgs::Point> SearchHelper::safetyPadding(const costmap_2d::Costmap2DROS* costmap_2dros,
+                                                              const bool use_padded_fp, const double safety_dist)
+
+{
+  std::vector<geometry_msgs::Point> footprint;
+  footprint = use_padded_fp ? costmap_2dros->getRobotFootprint() : costmap_2dros->getUnpaddedRobotFootprint();
+  costmap_2d::padFootprint(footprint, safety_dist);
+  return footprint;
+}
+
+SearchState SearchHelper::getFootprintState(const costmap_2d::Costmap2D* costmap_2d,
+                                            const std::vector<geometry_msgs::Point>& footprint,
+                                            const geometry_msgs::Pose2D& pose_2d)
+{
+  const auto cells_to_check =
+      FootprintHelper::getFootprintCells(pose_2d.x, pose_2d.y, pose_2d.theta, footprint, *costmap_2d, true);
+  if (cells_to_check.empty())
+  {
+    return { costmap_2d::NO_INFORMATION, SearchState::OUTSIDE };
+  }
+
+  // create a map of cells to avoid duplicates
+  std::unordered_map<int, Cell> cells_to_check_map;
+  for (const auto& cell : cells_to_check)
+  {
+    cells_to_check_map[costmap_2d->getIndex(cell.x, cell.y)] = cell;
+  }
+
+  unsigned char max_cost = 0;
+  for (const auto& [index, cell] : cells_to_check_map)
```

Comment:
You are not using index? Why is cells_to_check_map an unordered_map and not just a vector<Cell>

Reply:
The reason is in the comment above. I am creating an `unordered_map` to avoid duplicates. Best would be an `unordered_set`, but that requires a hash given by the `costmap` (index), so the `unordered_map` solves the problem here.